src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs

```diff
@@ -663,7 +663,7 @@ public BoundExpression Coalesce(BoundExpression left, BoundExpression right)
             Debug.Assert(left.Type.Equals(right.Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds));
             Debug.Assert(left.Type.IsReferenceType);
 
-            return new BoundNullCoalescingOperator(Syntax, left, right, Conversion.Identity, left.Type) { WasCompilerGenerated = true };
+            return new BoundNullCoalescingOperator(Syntax, left, right, Conversion.Identity, BoundNullCoalescingOperatorResultKind.LeftType, left.Type) { WasCompilerGenerated = true };
```

Comment:
That didn't seem obvious to me.
Seems we should assert that left is not nullable. Or make the kind a parameter for the caller to decide?

Reply:
We're asserting `left.Type.Equals(right.Type)` and `left.Type.IsReferenceType`.

---
In reply to: [173338326](https://github.com/dotnet/roslyn/pull/25358#discussion_r173338326) [](ancestors = 173338326)